pyrs990/pyrs990/annual_record.py

```diff
@@ -0,0 +1,150 @@
+from __future__ import annotations
+
+import csv
+from logging import getLogger
+from typing import Any, Dict, Iterator, List, NamedTuple, NewType, TextIO
+
+from .downloader import Downloader
+from .types import EIN
+
+_logger = getLogger(__name__)
+
+
+ANNUAL_FIELD_NAMES: List[str] = [
+    "RETURN_ID",
+    "FILING_TYPE",
+    "EIN",
+    "TAX_PERIOD",
+    "SUB_DATE",
+    "TAXPAYER_NAME",
+    "RETURN_TYPE",
+    "DLN",
+    "OBJECT_ID",
+]
+
+
+AnnualYear = NewType("AnnualYear", int)
+"""
+A filing year supported by the software. This will generally only
+include 2017-2019, at least for now.
+"""
+
+
+AnnualRecordDict = NewType("AnnualRecordDict", Dict[str, str])
+"""
+A dictionary presumed to have the keys listed in `ANNUAL_FIELD_NAMES`
+so that it can be used to create an instance of `AnnualRecord`.
+"""
+
+
+class AnnualRecord(NamedTuple):
+    """
+    A single record from an annual index (such as those hosted at
+    https://registry.opendata.aws/irs990/) that can be used to get
+    an object ID that will allow the actual filing to be retrieved
+    from the internet.
+
+    TODO: Audit the descriptions for correctness and completeness
+    """
+
+    @staticmethod
+    def from_csv(index_file: TextIO,) -> Iterator[AnnualRecord]:
+        for row in csv.DictReader(index_file):
+            record = AnnualRecord.from_dict(AnnualRecordDict(row))
+            yield record
+
+    @staticmethod
+    def from_dict(d: AnnualRecordDict,) -> AnnualRecord:
+        params: Dict[str, str] = {}
+        try:
+            for name in ANNUAL_FIELD_NAMES:
+                params[name.lower()] = d[name]
+        except KeyError as e:
+            _logger.error(f"malformed index row - '{e}' - '{d}'")
+            raise
+        return AnnualRecord(**params)
+
+    def __repr__(self):
+        return self.__str__()
+
+    def __str__(self):
+        return f"AnnualRecord(taxpayer_name='{self.taxpayer_name}', ein='{self.ein}')"
```

Comment:
Wanna add the rest of the things to this?

Reply:
I kept it short so it would be easy to read in a REPL. In theory the `__repr__` version is supposed to use correct Python syntax to instantiate a new instance of the thing, if possible, so maybe we should add the rest of the fields to that... but, on the other hand, I think the REPL uses `__repr__` by default, so then that defeats the point. I'm not sure how I feel.